cpp/src/strings/split/partition.cu

```diff
@@ -0,0 +1,248 @@
+//     col0  col1    col2
+// 0  abcde  ""      ""
+// 1  null   null    null
+// 2  a      _       bc_déf
+// 3  a      _      _bc
+// 4  ""     _       ab_cd
+// 5  ab     _       cd_
+//
+struct partition_fn
+{
+    column_device_view const d_strings;   // strings to split
+    string_view const d_delimiter;        // delimiter for split
+    string_index_pair* d_indices_left{};  // the three
+    string_index_pair* d_indices_delim{}; // output columns
+    string_index_pair* d_indices_right{}; // amigos
+
+    partition_fn( column_device_view const& d_strings, string_view const& d_delimiter,
+                  string_index_pair* d_indices_left = nullptr,
+                  string_index_pair* d_indices_delim = nullptr,
+                  string_index_pair* d_indices_right = nullptr)
+                : d_strings(d_strings), d_delimiter(d_delimiter),
+                  d_indices_left(d_indices_left), d_indices_delim(d_indices_delim), d_indices_right(d_indices_right) {}
+
+    __device__ void set_null_entries( size_type idx )
+    {
+        if( d_indices_left )
+        {
+            d_indices_left[idx] = string_index_pair{nullptr,0};
+            d_indices_delim[idx] = string_index_pair{nullptr,0};
+            d_indices_right[idx] = string_index_pair{nullptr,0};
+        }
+    }
+
+    __device__ size_type check_delimiter( size_type idx, string_view const& d_str, string_view::const_iterator& itr )
+    {
+        size_type offset = itr.byte_offset();
+        size_type pos = -1;
+        if( d_delimiter.empty() )
+        {
+            if( *itr <= ' ' )  // whitespace delimited
+                pos = offset;
+        }
+        else
+        {
+            auto bytes = std::min(d_str.size_bytes()-offset,d_delimiter.size_bytes());
+            if( d_delimiter.compare(d_str.data()+offset,bytes)==0 )
+                pos = offset;
+        }
+        if( pos >=0 ) // delimiter found, set results
+        {
+            d_indices_left[idx] = string_index_pair{d_str.data(),offset};
+            if( d_delimiter.empty() )
+            {
+                d_indices_delim[idx] = string_index_pair{d_str.data()+offset,1};
+                ++offset;
+            }
+            else
+            {
+                d_indices_delim[idx] = string_index_pair{d_delimiter.data(),d_delimiter.size_bytes()};
+                offset += d_delimiter.size_bytes();
+            }
+            d_indices_right[idx] = string_index_pair{d_str.data() + offset,d_str.size_bytes() - offset};
+        }
+        return pos;
+    }
+
+    __device__ void operator()(size_type idx)
+    {
+        if( d_strings.is_null(idx) )
+        {
+            set_null_entries(idx);
+            return;
+        }
+        string_view d_str = d_strings.element<string_view>(idx);
+        size_type pos = -1;
+        for( auto itr=d_str.begin(); (pos < 0) && (itr < d_str.end()); ++itr )
+            pos = check_delimiter(idx,d_str,itr);
+        if( pos < 0 ) // delimiter not found
```

Comment:
This is a classic `std::find_if`
```cpp
auto it = std::find_if(
    std::begin(d_str),
    std::end(d_str),
    [&] (auto itr) { return check_delimiter(idx,d_str,itr) >=  0; });

if (it == std::end(d_str))
```

Reply:
The `UnaryPredicate` parameter for `find_if` is character (`char_utf8` in the case of `d_str`) and not the iterator itself as implied by the `e` parameter. Also, the `d_delimiter` is a string and not necessarily a single character so I don't think `find_if` would work here.

Reply:
Motivation for the iterator? In many strings functions, I sometimes need to iterate through variable-width characters and once found, know the byte-position. The iterator provides a convenient mechanism for this. Also, computing byte-position from character position requires adding up bytes from the beginning of the string. The iterator keeps track of this position so this is not usually required.